staging/src/k8s.io/kube-scheduler/config/v1alpha2/types.go

```diff
@@ -202,3 +218,20 @@ type PluginConfig struct {
 	// Args defines the arguments passed to the plugins at the time of initialization. Args can have arbitrary structure.
 	Args runtime.RawExtension `json:"args,omitempty"`
 }
+
+func (c *PluginConfig) decodeNestedObjects(d runtime.Decoder) error {
+	gvk := SchemeGroupVersion.WithKind(c.Name + "Args")
+	obj, parsedGvk, err := d.Decode(c.Args.Raw, &gvk, nil)
+	if err != nil {
+		if runtime.IsNotRegisteredError(err) {
+			// Must be args for out-of-tree plugin.
+			return nil
```

Comment:
This doesn't distinguish between a NotRegistered error caused because no kind is registered for `gvk` (which we want to tolerate) and an error caused because their args have a kind/apiVersion specified that could not be decoded (which we don't want to tolerate).

It's ugly, but we could do a dry-run decode to our desired gvk to see if the type exists:

```go
gvk := SchemeGroupVersion.WithKind(c.Name + "Args")
if _, _, err := d.Decode([]byte(`{}`), &gvk, nil); err != nil && runtime.IsNotRegisteredError(err) {
  // no typed kind registered for this plugin's args
  return nil
}
```

then parse normally and treat any error as an actual error:
```go
obj, parsedGvk, err := d.Decode(c.Args.Raw, &gvk, nil)
if err != nil {
  return fmt.Errorf("decoding args for plugin %s: %w", s.Name, err)
}
```

and include more details if the kind doesn't match our expectation:
```go
if parsedGvk.GroupKind() != gvk.GroupKind() {
	return fmt.Errorf("args for plugin %s were not of type %s: %s", c.Name, gvk.GroupKind(), parsedGvk.GroupKind())
}
```

Reply:
make sure we have tests exercising the following:
1. no args (no error)
2. args without kind/apiVersion for a known plugin with a type registered (no error, get internal type)
3. args without kind/apiVersion for a custom plugin that does not have an args type registered (no error, get runtime.Unknown internally)
4. args with kind/apiVersion that match the expected GVK (no error, get internal type)
5. args with known kind/apiVersion that do not match the expected GVK (should get a groupkind mismatch error)
6. args with unknown kind/apiVersion (should get a NotRegistered error)

For the non-error cases, I would also then round-trip the resulting internal config back to v1alpha2 and encode it and make sure the arg types get converted/encoded correctly.

Reply:
Done